Draft: Shamir + BIP-39

danielstreit opened the pull request:
A proposal for a standardized algorithm for applying Shamir's Secret Sharing Scheme to BIP-39 mnemonics, where the shares are also BIP-39 mnemonics.

Reference implementation: https://github.com/danielstreit/shamir-bip39

Updated from #3445 to use a GF(2048) and avoid converting to and from hex values.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+## Test Cases
+
+[Test vectors](https://github.com/danielstreit/shamir-bip-39/blob/272333114ce03dfb5f4ed94f7ddb22e7cc0c93c4/test/vectors.ts)
```
No external links.  Test vectors should be inlined into this file, or included in `../assets/eip-3450/` and linked with a relative link.

danielstreit replied:
Removed. Will re-add test vectors as a file in assets shortly.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+## Reference Implementation
+
+https://github.com/danielstreit/shamir-bip39
```
Reference implementation should be inline, included as an asset, or this section can be removed (it is optional).

danielstreit replied:
Removed.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+are additional shares.
+
+A potential alternative specification might include a "digest" in the polynomial that could be used to validate the recovered secret. See, for example,
+[SLIP-0039](https://github.com/satoshilabs/slips/blob/master/slip-0039.md#shamirs-secret-sharing).
```
No external links.

danielstreit replied:
Removed.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+To mitigate leaking information, users can consider using an alias for ids or obscuring them in some manner.
+
+## References and Credits
+
+This specification leans heavily on prior work, including:
+
+- The [SLIP-0039 Specification](https://github.com/satoshilabs/slips/blob/master/slip-0039.md)
+- Grempe's [secrets.js](https://github.com/grempe/secrets.js) implementation of Shamir's Secret Sharing Scheme
+
```
```suggestion
```
This is not a valid EIP section.  Also, EIPs are CC0 which does not require attribution.  And of course, no external links.  😉

danielstreit replied:
Removed.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+   - See [Polynomial Interpolation](#polynomial-interpolation) above
+4. Convert _E_ to a mnemonic using each value to lookup a word by index in the word list
+
+## Test Cases
```
```suggestion
## Rationale
TBD

## Test Cases
```
A rationale section is required.  It should indicate why specific decisions were made such as why a certain constant was picked over another, or why one algorithm was chosen over others.

danielstreit replied:
Added.

MicahZoltu commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+eip: 3450
+title: Standardized Shamir Secret Sharing Scheme for BIP-39 Mnemonics
+author: Daniel Streit (@danielstreit)
+status: Draft
```
```suggestion
discussions-to: <URL>
status: Draft
```
`discussions-to` link is required.  We recommend a thread on https://ethereum-magicians.org/ but if that is not possible or you are unwilling to create one there then an issue in this repository can be used as an alternative.

danielstreit replied:
Added.

MicahZoltu reviewed (changes requested):
Discussions-to link and Rationale section are required and extraneous sections must be removed before merging as a draft.  The remaining issues *can* be addressed after merging as a Draft, but some will need to be addressed before moving this to Review stage.

danielstreit commented:
Thank you for the review @MicahZoltu! I apologize for breaking so many conventions but appreciate the gentle guidance. I will fix all these shortly.

danielstreit commented:
Update: There is ongoing discussion on this here: https://ethereum-magicians.org/t/erc-3450-standard-for-applying-shamirs-to-bip-39-mnemonics/5844/10.

I'll return to updating this EIP and applying the feedback once we bottom out on which field is best for this application.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+For a general overview of Shamir's Secret Sharing Scheme see:
+
+- [Wikipedia: Shamir's Secret Sharing](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)
+- [SLIP-0039: Shamir's Secret-Sharing for Mnemonic Codes](https://github.com/satoshilabs/slips/blob/master/slip-0039.md#shamirs-secret-sharing)
```
Fixed.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+There are two operations: Splitting a BIP-39 mnemonic into shares and recovering the original mnemonic with a subset of the shares.
+
+We use the [standard BIP-39 english word list](https://github.com/bitcoin/bips/blob/master/bip-0039/english.txt) with 2048 words, in a 0-indexed array.
```
Fixed. Included word list as an asset.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+We use the [standard BIP-39 english word list](https://github.com/bitcoin/bips/blob/master/bip-0039/english.txt) with 2048 words, in a 0-indexed array.
+
+Accordingly, we use GF(2048) (or GF(2<sup>11</sup>)) with primitive polynomial x<sup>11</sup> + x<sup>2</sup> + 1 (or 5) for all arithmetic.
```
I played around with this a bit, and it seems that neither `2**11` nor `2^11` are rendered as superscript for me. It looks like they are not part of the markdown spec either, though it appears that _some_ implementations honor them.

One option is to just use that notation and let it render without proper formatting (it is still legible). 

But, what would be the notation for subscript? I'm not familiar with a common, plain text alternative for this.

Agreed that HTML inside Markdown is not ideal, but to me, only using the `<sup>` and `<sub>` elements seems like a reasonable compromise and is well supported by most Markdown renderers I'm familiar with.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+Interpolation is used to recover the mnemonic. See [Recover Mnemonic](#recover-mnemonic) below.
+
+This follows the [SLIP-0039 specification for polynomial interpolation](https://github.com/satoshilabs/slips/blob/master/slip-0039.md#polynomial-interpolation) closely.
```
Removed this reference.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+Output: the value-vector (_f<sub>1</sub>_(_x_), ... , _f<sub>n</sub>_(_x_))
+
+![$f_l(x) = \sum_{i=1}^{K} y_i[l] \prod_{\underset{j \neq i}{j=1}}^{K} \frac{x - x_j}{x_i - x_j}$](../assets/eip-3450/lagrange.gif)
```
This isn't actually HTML, its the alt text of the formula image. The Markdown is `![alt text](./image/path)`. Here, the alt text is the LaTeX used to generate the formula image.

While python could be an option, the code is relatively long and, in my opinion, more difficult to follow.

The Barycentric form of Lagrange represented here is relatively common (implementations exist in many languages), so I don't actually expect many implementations of this spec to re-implement this function. Most _should_ use an existing math library for this.

So, to me, the mathematical representation is more generic and terse than including python source code.

See a python implementation here: https://github.com/scipy/scipy/blob/v1.6.2/scipy/interpolate/polyint.py#L497-L650.

danielstreit commented on EIPS/eip-3450.md:
```diff
@@ -0,0 +1,151 @@
+
+Output: the value-vector (_f<sub>1</sub>_(_x_), ... , _f<sub>n</sub>_(_x_))
+
+![$f_l(x) = \sum_{i=1}^{K} y_i[l] \prod_{\underset{j \neq i}{j=1}}^{K} \frac{x - x_j}{x_i - x_j}$](../assets/eip-3450/lagrange.gif)
```
While I don't disagree in most cases, mathematical formulas like this can be difficult to write out with more descriptive variables.

This particular formula is a form of Lagrange, and is relatively common. So, I feel like the mathematical notation here is reasonable.

danielstreit commented:
The CI is failing on code spell:

```
./assets/eip-3450/wordlist.txt:1712: sting ==> string
```
Here, the correct word _is_ "sting". Replacing it with "string" would break things.

How can I override this particular case?

danielstreit commented:
@MicahZoltu Thanks again for the review! I've addressed all your comments, either by updating the doc or replying to them directly.

I feel like this is ready for another review

MicahZoltu commented:
https://github.com/ethereum/EIPs/pull/3488 to fix the spellchecker.  Very bizarre failure since `sting` is a word.
